File: ruby/tests/file_io.rb
require "json"
require "tempfile"
require "time"

class FileIOTest
  attr_reader :name

  def initialize(size_mb = 50)
    @name = "File I/O"
    @size_mb = size_mb
    @min_free_mb = size_mb * 2
  end

  # Cross-platform free disk space
  def free_space_mb(dir)
    if Gem.win_platform?
      out = `wmic logicaldisk get caption,freespace /value`
      drive = dir[0].upcase

      out.split("\n").each do |line|
        if line.start_with?("Caption=#{drive}:")
          free = line.split("\n")[1].split("=")[1].to_i
          return free / (1024 * 1024)
        end
      end
      999_999
    else
      out = `df -m #{dir} 2>/dev/null`
      parts = out.split("\n")[1].split
      parts[3].to_i # free MB
    end
  rescue
    999_999
  end

  # Safe deletion for Windows handle issues
  def safe_delete(path)
    return unless File.exist?(path)

    5.times do
      begin
        File.delete(path)
        return true
      rescue Errno::EACCES
        sleep(0.05)
      end
    end

    File.delete(path) rescue nil
  end

  def run_once
    block = ("X" * 1024 * 1024).b

    # Tempfile we should NOT manually delete
    tempfile = Tempfile.new("cgv_test", Dir.tmpdir)
    @test_file = tempfile.path + "_data"  # <-- Use a separate manual file

    free_mb = free_space_mb(Dir.tmpdir)
    if free_mb < @min_free_mb
      raise "Insufficient disk space: need #{@min_free_mb}MB, got #{free_mb}MB"
    end

    begin
      # Write test file
      t0 = Process.clock_gettime(Process::CLOCK_MONOTONIC)
      File.open(@test_file, "wb") do |f|
        @size_mb.times { f.write(block) }
      end
      write_time = Process.clock_gettime(Process::CLOCK_MONOTONIC) - t0

      # Read test file
      t1 = Process.clock_gettime(Process::CLOCK_MONOTONIC)
      File.open(@test_file, "rb") do |f|
        f.read(1024 * 1024) while f.read(1024 * 1024)
      end
      read_time = Process.clock_gettime(Process::CLOCK_MONOTONIC) - t1

    ensure
      tempfile.close  # auto-deletes the temp file it created
      safe_delete(@test_file)
    end

    {
      "write_time_s" => write_time,
      "read_time_s"  => read_time,
      "write_MBps"   => @size_mb / write_time,
      "read_MBps"    => @size_mb / read_time
    }
  end

  def run(runs = 3, iterations = nil)
    results = Array.new(runs) { run_once }

    write_list = results.map { |r| r["write_MBps"] }.sort
    read_list  = results.map { |r| r["read_MBps"] }.sort
    mid = runs / 2

    {
      "name" => @name,
      "runs" => runs,
      "size_mb" => @size_mb,
      "median_write_MBps" => write_list[mid],
      "median_read_MBps"  => read_list[mid],
      "raw" => results
    }
  end
end

FileIoTest = FileIOTest

if __FILE__ == $0
  test = FileIOTest.new(ARGV[0] ? ARGV[0].to_i : 5)
  res = test.run
  File.write("results_ruby_fileio.json", JSON.pretty_generate(res))
  puts JSON.pretty_generate(res)
end

File: python/tests/file_io.py
import os
from pathlib import Path
import time
import json
import tempfile
import gc

class FileIOTest:
    def __init__(self, size_mb=50, name="FileIO"):
        self.name = name
        self.size_mb = size_mb
        self.test_file = os.path.join(tempfile.gettempdir(), "41e0a8c7d0.cgv")

    def run_once(self):
        data = ("X" * 1024 * 1024).encode("utf-8")  # 1 MB block

        # Write test
        t0 = time.perf_counter()
        with open(self.test_file, "wb") as f:
            for _ in range(self.size_mb):
                f.write(data)
        write_time = time.perf_counter() - t0

        # Read test
        t1 = time.perf_counter()
        with open(self.test_file, "rb") as f:
            while f.read(1024 * 1024):
                pass
        read_time = time.perf_counter() - t1

        try:
            os.remove(self.test_file)
        except OSError:
            pass

        return {
            "time_s": write_time + read_time,
            "ops": self.size_mb,
            "write_time_s": write_time,
            "read_time_s": read_time,
            "write_MBps": self.size_mb / write_time if write_time > 0 else 0,
            "read_MBps": self.size_mb / read_time if read_time > 0 else 0,
            "ops_per_sec": self.size_mb / (write_time + read_time)
        }

    def run(self, runs=None, iterations=None):
        if runs is None:
            runs = 3
        results = []
        for _ in range(runs):
            results.append(self.run_once())
            gc.collect()

        avg_write = sum(r["write_MBps"] for r in results) / runs
        avg_read = sum(r["read_MBps"] for r in results) / runs

        times = sorted(r["time_s"] for r in results)
        ops = sorted(r["ops_per_sec"] for r in results)
        mid = len(times) // 2

        return {
            "name": self.name,
            "runs": runs,
            "avg_write_MBps": avg_write,
            "avg_read_MBps": avg_read,
            "median_time_sec": times[mid],
            "median_ops_per_sec": ops[mid],
            "raw": results
        }